operator/reconciler.go

```diff
@@ -0,0 +1,276 @@
+package operator
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/utils/strings/slices"
+
+	"github.com/grafana/grafana-app-sdk/resource"
+)
+
+// ReconcileAction describes the action that triggered reconciliation.
+type ReconcileAction int
+
+const (
+	// ReconcileActionUnknown represents an Unknown ReconcileAction
+	ReconcileActionUnknown ReconcileAction = iota
+
+	// ReconcileActionCreated indicates that the resource to reconcile has been created.
+	// Note that this action may also be used on initial start-up of some informer-based implementations,
+	// such as the KubernetesBasedInformer. To instead receive Resync actions for these events,
+	// use the OpinionatedReconciler.
+	ReconcileActionCreated
+
+	// ReconcileActionUpdated indicates that the resource to reconcile has been updated.
+	ReconcileActionUpdated
+
+	// ReconcileActionDeleted indicates that the resource to reconcile has been deleted.
+	// Note that if the resource has Finalizers attached to it, a ReconcileActionUpdated will be used to indicate
+	// "tombstoning" of the resource where DeletionTimestamp is non-nil and Finalizers may only be removed.
+	// On completion of the actual delete from the API server once the Finalizers list is empty,
+	// a Delete reconcile action will be triggered.
+	ReconcileActionDeleted
+
+	// ReconcileActionResynced indicates a periodic or initial re-sync of existing resources in the API server.
+	// Note that not all implementations support this action (KubernetesBasedInformer will only trigger Created,
+	// Updated, and Deleted actions. You can use OpinionatedReconciler to introduce Resync events on start instead
+	// of Add events).
+	ReconcileActionResynced
+)
+
+// ReconcileRequest contains the action which took place, and a snapshot of the object at that point in time.
+// The Object in the ReconcileRequest is not guaranteed to be the current state of the object in-storage,
+// as other actions may have taken place subsequently.
+//
+// Controllers such as InformerController contain logic to dequeue ReconcileRequests if subsequent actions
+// are received for the same object.
+type ReconcileRequest struct {
+	// Action is the action that triggered this ReconcileRequest
+	Action ReconcileAction
+	// Object is the object art the time of the received action
+	Object resource.Object
+	// State is a user-defined map of state values that can be provided on retried ReconcileRequests.
+	// See State in ReconcileResult. It will always be nil on an initial Reconcile call,
+	// and will only be non-nil if a prior Reconcile call with this ReconcileRequest returned a State
+	// in its ReconcileResult alongside either a RequeueAfter or an error.
+	State map[string]any
```

Comment:
Not sure if it's worth adding this for extra complexity. Normally it's expected that the reconciler will handle storing intermediate state itself, by e.g. updating the object's status in the API.

Reply:
Yeah, this was mainly me rabbit-holing myself in edge cases where the reconciler successfully calls a third-party service, but then calls to update the object's status subresource fail for a transient reason (i.e. 5xx), so need to be retried in some exponential backoff way. It _does_ introduce a lot of complexity, so I think we can remove it if it doesn't get any usage.

Reply:
I generally prefer to only err on the side of not adding things which might not end up being used, but I don't have a strong objection in this particular case.